singularity-eos/closure/mixed_cell_models.hpp

```diff
@@ -272,450 +129,636 @@ PORTABLE_INLINE_FUNCTION bool solve_Ax_b(Real *a, Real *b) {
-PORTABLE_INLINE_FUNCTION bool
-pte_closure_flag_with_line_impl(EOSIndexer &&eoss, const Real Volume, const Real TotalSIE,
-                                ConstRealIndexer &&ComponentMasses,
-                                RealIndexer &&ComponentVolumes,
-                                RealIndexer &&ComponentEnergies, LambdaIndexer &&lambdas,
-                                const OFFSETTER ofst, int &iter) {
-  using namespace mix_params;
+struct NullPtrIndexer {
+  PORTABLE_INLINE_FUNCTION Real *operator[](const int i) { return nullptr; }
+};
 
-  Real TotalMass = 0.0, Vsum = 0.0, Esum = 0.0;
-  Real Densities[nmat], Pressures[nmat], Temperatures[nmat], dpdr[nmat];
-  Real dpde[nmat], dtdr[nmat], dtde[nmat], b[2 * nmat], A[4 * nmat * nmat];
-  Real rtemp[nmat], sietemp[nmat];
-  // First, normalize the energies and volumes.
-  // This may or may not be close to reality since SIE isn't strictly positive.
-  for (int i = 0; i < nmat; ++i)
-    TotalMass += ComponentMasses[i];
-  bool iszero = true;
-  for (int i = 0; i < nmat; ++i)
-    iszero &= ComponentVolumes[i] <= 1.0e-16;
-  if (iszero)
-    for (int i = 0; i < nmat; ++i)
-      ComponentVolumes[i] = Volume * ComponentMasses[i] / TotalMass;
-  for (int i = 0; i < nmat; ++i)
-    Vsum += ComponentVolumes[i];
-  for (int i = 0; i < nmat; ++i)
-    ComponentVolumes[i] *=
-        Volume / Vsum; // Are compilers smart enough to calculate this RHS once?
-  for (int i = 0; i < nmat; ++i)
-    Esum += ComponentEnergies[i] * ComponentMasses[i];
-  if (std::abs(Esum) <
-      0.0001) { // Then we punt and just distribute based on mass fractions
-    for (int i = 0; i < nmat; ++i)
-      ComponentEnergies[i] = TotalSIE * TotalMass / ComponentMasses[i];
-  } else { // Otherwise, we scale
-    for (int i = 0; i < nmat; ++i)
-      ComponentEnergies[i] *= (TotalSIE * TotalMass / Esum);
+class CacheAccessor {
+ public:
+  PORTABLE_INLINE_FUNCTION
+  CacheAccessor() = default;
+  PORTABLE_INLINE_FUNCTION
+  explicit CacheAccessor(Real *scr) : cache_(scr) {}
+  PORTABLE_FORCEINLINE_FUNCTION
+  Real * operator[](const int m) const {
+    return cache_ + m*MAX_NUM_LAMBDAS;
   }
+ private:
+  Real *cache_;
+};
 
-  // This is the main loop
-  constexpr const int pte_max_iter = nmat * pte_max_iter_per_mat;
-  for (iter = 0; iter < pte_max_iter; ++iter) {
-    // Call the EOSs
-    for (int mat = 0; mat < nmat; ++mat) {
-      Densities[mat] = ComponentMasses[mat] / ComponentVolumes[mat];
-      eoss[ofst(mat)].PTofRE(Densities[mat], ComponentEnergies[mat], lambdas[mat],
-                             Pressures[mat], Temperatures[mat], dpdr[mat], dpde[mat],
-                             dtdr[mat], dtde[mat]);
+template <typename EOSIndexer, typename RealIndexer>
+class PTESolverBase {
+ public:
+  PTESolverBase() = delete;
+  PORTABLE_INLINE_FUNCTION int Nmat() const { return nmat; }
+  PORTABLE_INLINE_FUNCTION int &Niter() { return niter; }
+  PORTABLE_INLINE_FUNCTION
+  virtual void Fixup() const {
```

Comment:
I had imagined there might be other solvers with needs beyond normalizing the volume fractions.  Actually, for the `PTESolverRhoU` solver, we might want to be normalizing the energies as well.  Not in there yet, but I was also thinking of adding another solver that could act as a fallback if PTE failed.  Could, for example, try to just get pressure equilibrium and toss out temperature equilibrium.  One formulation I'm thinking of wouldn't even update volume fractions and so wouldn't need to normalize them.  So, in short, I call it `Fixup` because it's supposed to be a hook to apply solver specific special sauce that enforces certain things that would vary from solver to solver.  Make sense?

Reply:
Ah gotcha. That makes sense. Could you add a summary of that in a comment?

Reply:
yeah, will do.

Reply:
Yeah I saw what you were doing here @jdolence and didn't love the name `Fixup`, but I couldn't think of a better one.